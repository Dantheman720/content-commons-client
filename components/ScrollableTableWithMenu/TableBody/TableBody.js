import './TableBody.scss';
import { Loader, Table } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import PropTypes from 'prop-types';
import { graphql } from 'react-apollo';
import React, { useEffect, useState } from 'react';
import update from 'immutability-helper';
import isEqual from 'lodash/isEqual';
import TableRow from 'components/ScrollableTableWithMenu/TableRow/TableRow';
import orderBy from 'lodash/orderBy';
import { PROJECT_STATUS_CHANGE_SUBSCRIPTION } from 'lib/graphql/queries/common';
import {
  TEAM_VIDEO_PROJECTS_QUERY,
  // DASHBOARD_PROJECTS_QUERY - queries videoProjects, packages, documentFiles
} from 'lib/graphql/queries/dashboard';

// TEMP
import { packageMocks, documentFileMocks } from './pressMocks';

const teamPackages = packageMocks[0].result.data;
const teamDocumentFiles = documentFileMocks[0].result.data;

const getLangTaxonomies = ( array, locale = 'en-us' ) => {
  if ( !Array.isArray( array ) || !array.length ) return '';
  return (
    array.map( tax => (
      tax.translations
        .find( translation => translation.language.locale === locale )
        .name
    ) ).join( ', ' )
  );
};

const normalizeTypesData = type => {
  const projectTitle = () => {
    if ( type.__typename === 'VideoProject' ) return type.projectTitle;
    if ( type.__typename === 'Package' ) return type.title;
    if ( type.__typename === 'DocumentFile' ) return type.filename;
    return '';
  };

  const thumbnail = () => {
    if ( !type.thumbnail ) return {};
    return ( {
      signedUrl: type.thumbnails && type.thumbnails.length ? type.thumbnails[0].signedUrl : '',
      alt: type.thumbnails && type.thumbnails.length ? type.thumbnails[0].alt : ''
    } );
  };

  const normalizedTypeData = Object.create( {}, {
    __typename: { value: type.__typename },
    id: { value: type.id },
    createdAt: { value: type.createdAt },
    updatedAt: { value: type.updatedAt },
    projectTitle: { value: projectTitle() },
    author: { value: `${type.author ? type.author.firstName : ''} ${type.author ? type.author.lastName : ''}` },
    team: { value: type.team ? type.team.name : '' },
    status: { value: type.status || '' },
    visibility: { value: type.visibility },
    thumbnail: { value: thumbnail() },
    categories: { value: getLangTaxonomies( type.categories ) }
  } );

  return normalizedTypeData;
};

const normalizeDashboardData = types => {
  const normalizedProjects = [];
  types.forEach( type => normalizedProjects.push( normalizeTypesData( type ) ) );
  return normalizedProjects;
};

const getTypesData = types => {
  let typesData = [];
  const filterTypesData = types.filter( type => type !== null );
  typesData = typesData.concat( ...filterTypesData );
  return typesData;
};

const TableBody = props => {
  const {
    searchTerm,
    selectedItems,
    tableHeaders,
    toggleItemSelection,
    variables,
    projectTab,
    teamVideoProjects: {
      loading, error, videoProjects
    },
    videoProjectIds,
    subscribeToStatuses
  } = props;

  // TEMP - add mock data w/ videoProjects query result
  const allProjectTypesData = getTypesData( [
    videoProjects,
    teamDocumentFiles,
    teamPackages
  ] );

  const [statusSubscription, setStatusSubscription] = useState();

  useEffect( () => {
    if ( statusSubscription ) {
      // Do not resubscribe if the IDs did not change
      if ( isEqual( statusSubscription.ids, videoProjectIds ) ) {
        return;
      }
      statusSubscription.unsub();
    }
    if ( videoProjectIds.length < 1 ) {
      setStatusSubscription( null );
      return;
    }
    const subscription = {
      ids: videoProjectIds,
      unsub: subscribeToStatuses()
    };
    setStatusSubscription( subscription );
    return () => {
      subscription.unsub();
    };
  }, [videoProjectIds.join( ',' )] );

  if ( loading ) {
    return (
      <Table.Body>
        <Table.Row>
          <Table.Cell>
            <Loader active inline size="small" />
            <span style={ { marginLeft: '0.5em', fontSize: '1.5em' } }>
              Loading...
            </span>
          </Table.Cell>
        </Table.Row>
      </Table.Body>
    );
  }
  if ( error ) {
    return (
      <Table.Body>
        <Table.Row>
          <Table.Cell>
            <ApolloError error={ error } />
          </Table.Cell>
        </Table.Row>
      </Table.Body>
    );
  }

  // if ( !videoProjects ) return null;
  if ( !allProjectTypesData ) return null;

  // if ( searchTerm && !videoProjects.length ) {
  if ( searchTerm && !allProjectTypesData.length ) {
    return (
      <Table.Body>
        <Table.Row>
          <Table.Cell>
            No results for &ldquo;{ searchTerm }&rdquo;
          </Table.Cell>
        </Table.Row>
      </Table.Body>
    );
  }

  // if ( !videoProjects.length ) {
  if ( !allProjectTypesData.length ) {
    return (
      <Table.Body>
        <Table.Row>
          <Table.Cell>No projects</Table.Cell>
        </Table.Row>
      </Table.Body>
    );
  }

  // Sort data by clicked column & direction
  // Default sort by createdAt & DESC
  const direction = props.direction ? `${props.direction === 'ascending' ? 'asc' : 'desc'}` : 'desc';

  const tableData = orderBy(
    normalizeDashboardData( allProjectTypesData ),
    tableDatum => {
      let { column } = props;
      if ( !column ) column = 'createdAt';
      // Format table data for case insensitive sorting
      const formattedTableDatum = tableDatum[column].toString().toLowerCase();
      return formattedTableDatum;
    },
    [direction]
  );

  // skip & first query vars are used as start/end slice() params to paginate tableData on client
  const { skip, first } = variables;
  const paginatedTableData = tableData.slice( skip, skip + first );

  return (
    <Table.Body className="projects">
      { paginatedTableData.map( d => (
        <TableRow
          key={ d.id }
          d={ d }
          selectedItems={ selectedItems }
          tableHeaders={ tableHeaders }
          toggleItemSelection={ toggleItemSelection }
          projectTab={ projectTab }
        />
      ) ) }
    </Table.Body>
  );
};

TableBody.propTypes = {
  column: PropTypes.string,
  teamVideoProjects: PropTypes.object,
  searchTerm: PropTypes.string,
  selectedItems: PropTypes.object,
  tableHeaders: PropTypes.array,
  toggleItemSelection: PropTypes.func,
  variables: PropTypes.object,
  direction: PropTypes.string,
  projectTab: PropTypes.string,
  videoProjectIds: PropTypes.array,
  subscribeToStatuses: PropTypes.func
};

const updateProjectStatus = ( prev, { subscriptionData: { data: { projectStatusChange } } } ) => {
  if ( !projectStatusChange ) {
    return prev;
  }
  const projectIndex = prev.videoProjects.findIndex( p => p.id === projectStatusChange.id );
  if ( projectIndex === -1 ) {
    return prev;
  }
  // Using immutability helper in order to ensure that React will rerender after the status change
  return update( prev, { videoProjects: { [projectIndex]: { status: { $set: projectStatusChange.status } } } } );
};

const teamProjectsQuery = graphql( TEAM_VIDEO_PROJECTS_QUERY, {
  name: 'teamVideoProjects',
  props: ( { teamVideoProjects } ) => {
    const { videoProjects, subscribeToMore } = teamVideoProjects;
    const videoProjectIds = videoProjects ? videoProjects.map( p => p.id ) : [];
    return {
      teamVideoProjects,
      videoProjectIds,
      subscribeToStatuses: () => subscribeToMore( {
        document: PROJECT_STATUS_CHANGE_SUBSCRIPTION,
        variables: { ids: videoProjectIds },
        updateQuery: updateProjectStatus
      } )
    };
  },
  options: props => ( {
    variables: { ...props.variables },
    notifyOnNetworkStatusChange: true,
    fetchPolicy: 'cache-and-network'
  } )
} );

// const dashboardProjectsQuery = graphql( DASHBOARD_PROJECTS_QUERY, {
//   name: 'dashboardProjects',
//   props: ( { dashboardProjects } ) => {
//     const {
//       subscribeToMore,
//       documentFiles,
//       packages,
//       videoProjects
//     } = dashboardProjects;
//     const allDashboardProjects = getTypesData( [ documentFiles, packages, videoProjects ] );
//     const dashboardProjectIDs = allDashboardProjects ? allDashboardProjects.map( p => p.id ) : [];
//     return {
//       dashboardProjects,
//       dashboardProjectIDs,
//       subscribeToStatuses: () => subscribeToMore( {
//         document: PROJECT_STATUS_CHANGE_SUBSCRIPTION,
//         variables: { ids: dashboardProjectIDs },
//         updateQuery: ( prev, { subscriptionData: { data: { projectStatusChange } } } ) => {
//           if ( !projectStatusChange ) {
//             return prev;
//           }
//           const projectIndex = prev.dashboardProjects.findIndex( p => p.id === projectStatusChange.id );
//           if ( projectIndex === -1 ) {
//             return prev;
//           }
//           return update( prev, { dashboardProjects: { [projectIndex]: { status: { $set: projectStatusChange.status } } } } );
//         }
//       } )
//     };
//   },
//   options: props => ( {
//     variables: { ...props.variables },
//     notifyOnNetworkStatusChange: true,
//     fetchPolicy: 'cache-and-network'
//   } )
// } );

export default teamProjectsQuery( TableBody );

const TableBodyRaw = TableBody;
export {
  TEAM_VIDEO_PROJECTS_QUERY, updateProjectStatus, teamProjectsQuery, TableBodyRaw
};
