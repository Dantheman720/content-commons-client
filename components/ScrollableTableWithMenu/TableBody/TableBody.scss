.myProjects {
  .items_table_item {
    vertical-align: top;
  }
}
