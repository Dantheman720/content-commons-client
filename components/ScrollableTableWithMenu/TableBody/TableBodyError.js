import React from 'react';
import PropTypes from 'prop-types';
import ApolloError from 'components/errors/ApolloError';
import { Table } from 'semantic-ui-react';

const TableBodyError = props => (
  <Table.Body>
    <Table.Row>
      <Table.Cell>
        <ApolloError error={ props.error } />
      </Table.Cell>
    </Table.Row>
  </Table.Body>
);

TableBodyError.propTypes = {
  error: PropTypes.object
};

export default TableBodyError;
