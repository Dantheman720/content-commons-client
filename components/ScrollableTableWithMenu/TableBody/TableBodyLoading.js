import React from 'react';
import { Loader, Table } from 'semantic-ui-react';

const TableBodyLoading = () => (
  <Table.Body>
    <Table.Row>
      <Table.Cell>
        <Loader active inline size="small" />
        <span style={ { marginLeft: '0.5em', fontSize: '1.5em' } }>
          Loading...
        </span>
      </Table.Cell>
    </Table.Row>
  </Table.Body>
);

export default TableBodyLoading;
