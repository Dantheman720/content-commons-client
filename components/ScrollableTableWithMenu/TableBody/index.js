import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { Loader, Table } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import MyProjectPrimaryCol from 'components/admin/Dashboard/MyProjects/MyProjectPrimaryCol';
import TableMobileDataToggleIcon from 'components/ScrollableTableWithMenu/TableMobileDataToggleIcon';
import { formatDate } from 'lib/utils';
import './TableBody.scss';

const TEAM_VIDEO_PROJECTS_QUERY = gql`
  query VideoProjectsByTeam(
    $team: String!, $first: Int, $skip: Int, $searchTerm: String
  ) {
    videoProjects(
      where: {
        AND: [
          { team: { name: $team } },
          {
            OR: [
              { projectTitle_contains: $searchTerm },
              { descPublic_contains: $searchTerm },
              { descInternal_contains: $searchTerm },
              { projectType_contains: $searchTerm },
              {
                categories_some: {
                  translations_some: { name_contains: $searchTerm }
                }
              },
              {
                author: {
                  OR: [
                    { firstName_contains: $searchTerm },
                    { lastName_contains: $searchTerm },
                    { email_contains: $searchTerm }
                  ]
                }
              }
            ]
          }
        ]
      },
      first: $first,
      skip: $skip
     ) {
      id
      createdAt
      updatedAt
      team {
        name
        organization
      }
      author {
        firstName
        lastName
      }
      projectTitle
      status
      visibility
      thumbnails {
        url
        alt
        caption
        filename
        filetype
        dimensions {
          width
          height
        }                
      }      
    }
  }
`;

const normalizeData = videoProjects => {
  const normalizedVideoProjects = [];

  videoProjects.forEach( videoProject => {
    const normalizedProject = Object.create( {}, {
      id: { value: videoProject.id },
      createdAt: { value: formatDate( videoProject.createdAt ) },
      updatedAt: { value: formatDate( videoProject.updatedAt ) },
      projectTitle: { value: videoProject.projectTitle },
      author: { value: `${videoProject.author ? videoProject.author.firstName : ''} ${videoProject.author ? videoProject.author.lastName : ''}` },
      team: { value: videoProject.team.name },
      status: { value: videoProject.status },
      visibility: { value: videoProject.visibility },
      thumbnail: {
        value: {
          url: videoProject.thumbnails[0].url,
          alt: videoProject.thumbnails[0].alt
        }
      }
    } );
    normalizedVideoProjects.push( normalizedProject );
  } );

  return normalizedVideoProjects;
};

const TableBody = props => {
  const {
    searchTerm,
    selectedItems,
    tableHeaders,
    toggleItemSelection,
    variables,
    windowWidth
  } = props;

  return (
    <Query
      query={ TEAM_VIDEO_PROJECTS_QUERY }
      variables={ { ...variables } }
    >
      { ( { loading, error, data: { videoProjects } } ) => {
        if ( loading ) {
          return (
            <Table.Body>
              <Table.Row>
                <Table.Cell>
                  <Loader active inline size="small" />
                  <span style={ { marginLeft: '0.5em', fontSize: '1.5em' } }>
                    Loading...
                  </span>
                </Table.Cell>
              </Table.Row>
            </Table.Body>
          );
        }
        if ( error ) {
          return (
            <Table.Body>
              <Table.Row>
                <Table.Cell>
                  <ApolloError error={ error } />
                </Table.Cell>
              </Table.Row>
            </Table.Body>
          );
        }
        if ( !videoProjects ) return null;

        if ( searchTerm && !videoProjects.length ) {
          return (
            <Table.Body>
              <Table.Row>
                <Table.Cell>
                  No results for &ldquo;{ searchTerm }&rdquo;
                </Table.Cell>
              </Table.Row>
            </Table.Body>
          );
        }

        const tableData = normalizeData( videoProjects );

        return (
          <Table.Body className="myProjects">
            { tableData.map( d => (
              <Table.Row
                key={ d.id }
                className={ d.isNew ? 'myProjects_newItem' : '' }
              >
                { tableHeaders.map( ( header, i ) => (
                  <Table.Cell
                    data-header={ header.label }
                    key={ `${d.id}_${header.name}` }
                    className="items_table_item"
                  >
                    { i === 0
                      ? (
                        // Table must include .primary_col div for fixed column
                        <Fragment>
                          <div className="primary_col">
                            <MyProjectPrimaryCol
                              d={ d }
                              header={ header }
                              selectedItems={ selectedItems }
                              toggleItemSelection={ toggleItemSelection }
                            />
                          </div>
                          { windowWidth && <TableMobileDataToggleIcon /> }
                        </Fragment>
                      )
                      : (
                        <Fragment>
                          <span>
                            <div className="items_table_mobileHeader">{ header.label }</div>
                            { d[header.name] }
                          </span>
                          <br />
                          { header.label === 'MODIFIED'
                            ? <span>{ d.status }</span>
                            : null }
                        </Fragment>
                      ) }
                  </Table.Cell>
                ) ) }
              </Table.Row>
            ) ) }
          </Table.Body>
        );
      } }
    </Query>
  );
};

TableBody.propTypes = {
  searchTerm: PropTypes.string,
  selectedItems: PropTypes.object,
  tableHeaders: PropTypes.array,
  toggleItemSelection: PropTypes.func,
  variables: PropTypes.object,
  windowWidth: PropTypes.oneOfType( [
    PropTypes.string,
    PropTypes.number
  ] )
};

export default TableBody;
export { TEAM_VIDEO_PROJECTS_QUERY };
