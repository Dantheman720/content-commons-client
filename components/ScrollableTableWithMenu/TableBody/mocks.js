import { TEAM_VIDEO_PROJECTS_QUERY } from 'components/ScrollableTableWithMenu/TableBody/TableBody';
import { PROJECT_STATUS_CHANGE_SUBSCRIPTION } from 'lib/graphql/queries/common';

export const videoProjects = [
  {
    id: 'd137',
    createdAt: '2019-05-09T18:33:03.368Z',
    updatedAt: '2019-05-09T18:33:03.368Z',
    team: {
      id: 't888',
      name: 'IIP Video Production',
      organization: 'Department of State'
    },
    author: {
      id: 'a928',
      firstName: 'Jane',
      lastName: 'Doe'
    },
    projectTitle: 'Test Title 1',
    status: 'PUBLISHED',
    visibility: 'INTERNAL',
    thumbnails: {
      id: 't34',
      url: 'https://thumbnailurl.com',
      signedUrl: 'https://thumbnailurl.com',
      alt: 'some alt text',
    },
    categories: [
      {
        id: '38s',
        translations: [
          {
            id: '832',
            name: 'about america',
            language: {
              id: 'en23',
              locale: 'en-us'
            }
          }
        ]
      }
    ]
  },
  {
    id: 'z132',
    createdAt: '2019-05-12T18:33:03.368Z',
    updatedAt: '2019-05-12T18:33:03.368Z',
    team: {
      id: 't888',
      name: 'IIP Video Production',
      organization: 'Department of State'
    },
    author: {
      id: 'a287',
      firstName: 'Joe',
      lastName: 'Schmoe'
    },
    projectTitle: 'Test Title 2',
    status: 'PUBLISHED',
    visibility: 'INTERNAL',
    thumbnails: {
      id: 't34',
      url: 'https://thumbnailurl.com',
      signedUrl: 'https://thumbnailurl.com',
      alt: 'some alt text',
    },
    categories: [
      {
        id: '38s',
        translations: [
          {
            id: '832',
            name: 'about america',
            language: {
              id: 'en23',
              locale: 'en-us'
            }
          }
        ]
      }
    ]
  }
];

export const videoProjectIds = videoProjects.map( p => p.id );

export const propVariables = {
  team: 'IIP Video Production',
  searchTerm: '',
  first: 4,
  skip: 0
};

export const mocks = [
  {
    request: {
      query: TEAM_VIDEO_PROJECTS_QUERY,
      variables: { ...propVariables }
    },
    result: {
      data: { videoProjects },
      subscribeToStatuses: jest.fn()
    }
  },
  {
    request: {
      query: PROJECT_STATUS_CHANGE_SUBSCRIPTION,
      variables: { ids: videoProjectIds }
    },
    result: {
      data: {
        projectStatusChange: {
          id: 'd137',
          status: 'DRAFT',
          error: null
        }
      }
    }
  }
];
