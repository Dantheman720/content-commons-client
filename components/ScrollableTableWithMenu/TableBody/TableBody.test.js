import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { Loader, Table } from 'semantic-ui-react';
import ApolloError from 'components/errors/ApolloError';
import { PROJECT_STATUS_CHANGE_SUBSCRIPTION } from 'lib/graphql/queries/common';
import TableBody, { TEAM_VIDEO_PROJECTS_QUERY } from './TableBody';

/**
 * Use custom element to avoid "incorrect casing" error msg
 * @see https://jestjs.io/docs/en/tutorial-react.html#snapshot-testing-with-mocks-enzyme-and-react-16
 */
// jest.mock( 'next-server/dynamic', () => () => 'VideoDetailsPopup' );
jest.mock( 'next-server/dynamic', () => () => 'video-details-popup' );

// Mock DetailsPopup component since it's tested elsewhere
jest.mock( 'components/admin/Dashboard/TeamProjects/DetailsPopup/DetailsPopup', () => () => 'DetailsPopup' );

const videoProjects = [
  {
    id: 'd137',
    createdAt: '2019-05-09T18:33:03.368Z',
    updatedAt: '2019-05-09T18:33:03.368Z',
    team: {
      id: 't888',
      name: 'IIP Video Production',
      organization: 'Department of State'
    },
    author: {
      id: 'a928',
      firstName: 'Jane',
      lastName: 'Doe'
    },
    projectTitle: 'Test Title 1',
    status: 'PUBLISHED',
    visibility: 'INTERNAL',
    thumbnails: {
      id: 't34',
      url: 'https://thumbnailurl.com',
      alt: 'some alt text',
    },
    categories: [
      {
        id: '38s',
        translations: [
          {
            id: '832',
            name: 'about america',
            language: {
              id: 'en23',
              locale: 'en-us'
            }
          }
        ]
      }
    ]
  },
  {
    id: 'z132',
    createdAt: '2019-05-12T18:33:03.368Z',
    updatedAt: '2019-05-12T18:33:03.368Z',
    team: {
      id: 't888',
      name: 'IIP Video Production',
      organization: 'Department of State'
    },
    author: {
      id: 'a287',
      firstName: 'Joe',
      lastName: 'Schmoe'
    },
    projectTitle: 'Test Title 2',
    status: 'PUBLISHED',
    visibility: 'INTERNAL',
    thumbnails: {
      id: 't34',
      url: 'https://thumbnailurl.com',
      alt: 'some alt text',
    },
    categories: [
      {
        id: '38s',
        translations: [
          {
            id: '832',
            name: 'about america',
            language: {
              id: 'en23',
              locale: 'en-us'
            }
          }
        ]
      }
    ]
  }
];

const videoProjectIds = videoProjects.map( p => p.id );

const props = {
  searchTerm: '',
  selectedItems: new Map(),
  tableHeaders: [
    { name: 'projectTitle', label: 'PROJECT TITLE' },
    { name: 'visibility', label: 'VISIBILITY' },
    { name: 'updatedAt', label: 'MODIFIED' },
    { name: 'team', label: 'TEAM' }
  ],
  toggleItemSelection: jest.fn(),
  variables: {
    team: 'IIP Video Production',
    searchTerm: '',
    first: 4,
    skip: 0
  },
  direction: 'descending',
  projectTab: 'teamProjects',
  videoProjectIds
};

const mocks = [
  {
    request: {
      query: TEAM_VIDEO_PROJECTS_QUERY,
      variables: { ...props.variables }
    },
    result: {
      data: { videoProjects }
    }
  },
  {
    request: {
      query: PROJECT_STATUS_CHANGE_SUBSCRIPTION,
      variables: { ids: videoProjectIds }
    },
    result: {
      data: {
        projectStatusChange: {
          id: 'd137',
          status: 'DRAFT',
          error: null
        }
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <Table>
      <TableBody { ...props } />
    </Table>
  </MockedProvider>
);

describe( '<TableBody />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const tableBody = wrapper.find( TableBody );
    const loader = <Loader active inline size="small" />;

    expect( tableBody.exists() ).toEqual( true );
    expect( tableBody.contains( loader ) ).toEqual( true );
    expect( tableBody.contains( 'Loading...' ) ).toEqual( true );
  } );

  it( 'renders error message if an error is returned', async () => {
    const errorMocks = [
      {
        request: {
          query: TEAM_VIDEO_PROJECTS_QUERY,
          variables: { ...props.variables }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <Table>
          <TableBody { ...props } />
        </Table>
      </MockedProvider>
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const tableBody = wrapper.find( TableBody );
    const errorComponent = tableBody.find( ApolloError );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders null if videoProjects is null', async () => {
    const nullMocks = [
      {
        request: {
          query: TEAM_VIDEO_PROJECTS_QUERY,
          variables: { ...props.variables }
        },
        result: {
          data: { videoProjects: null }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ nullMocks } addTypename={ false }>
        <Table>
          <TableBody { ...props } />
        </Table>
      </MockedProvider>
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const tableBody = wrapper.find( TableBody );

    expect( tableBody.html() ).toEqual( null );
  } );

  it( 'renders a "No projects" message if there are no video projects', async () => {
    const emptyMocks = [
      {
        request: {
          query: TEAM_VIDEO_PROJECTS_QUERY,
          variables: { ...props.variables }
        },
        result: {
          data: { videoProjects: [] }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ emptyMocks } addTypename={ false }>
        <Table>
          <TableBody { ...props } />
        </Table>
      </MockedProvider>
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const tableBody = wrapper.find( TableBody );

    expect( tableBody.contains( 'No projects' ) ).toEqual( true );
  } );

  it( 'renders a "No results" message if there are no search results', async () => {
    const newSearchTerm = 'new term';

    const newProps = {
      ...{ ...props },
      ...{
        searchTerm: newSearchTerm,
        variables: {
          ...props.variables,
          searchTerm: newSearchTerm
        }
      }
    };

    const noSearchResultsMocks = [
      {
        request: {
          query: TEAM_VIDEO_PROJECTS_QUERY,
          variables: { ...newProps.variables }
        },
        result: {
          data: { videoProjects: [] }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ noSearchResultsMocks } addTypename={ false }>
        <Table>
          <TableBody { ...newProps } />
        </Table>
      </MockedProvider>
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const tableBody = wrapper.find( TableBody );
    const noResultsMsg = (
      <Table.Cell>
        No results for &ldquo;{ newProps.searchTerm }&rdquo;
      </Table.Cell>
    );

    expect( tableBody.contains( noResultsMsg ) ).toEqual( true );
  } );

  it( 'renders the correct table row(s)', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const tableBody = wrapper.find( TableBody );
    const tableRows = tableBody.find( 'tr' );
    const rowCount = tableRows.length;
    const expectedRowCount = mocks[0].result.data.videoProjects.length;

    expect( rowCount ).toEqual( expectedRowCount );
    expect( toJSON( tableRows ) ).toMatchSnapshot();
  } );

  it( 'subscribes to status updates', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();
    // const tableBody = wrapper.find( 'TableBody' );
    const tableBody = wrapper.findWhere( c => c.props().subscribeToStatuses !== undefined );
    // TODO: Find a way to test this function added by graphql HOC props option
  } );
} );
