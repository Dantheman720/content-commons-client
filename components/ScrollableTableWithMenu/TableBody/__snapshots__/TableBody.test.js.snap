// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableBody /> renders the correct table row(s) 1`] = `
Array [
  <tr
    className=""
  >
    <TableCell
      as="td"
      className="items_table_item"
      data-header="PROJECT TITLE"
      key="z132_projectTitle"
    >
      <td
        className="items_table_item"
        data-header="PROJECT TITLE"
      >
        <div
          className="primary_col"
        >
          <video-details-popup
            d={Object {}}
            header={
              Object {
                "label": "PROJECT TITLE",
                "name": "projectTitle",
              }
            }
            selectedItems={Array []}
            team={
              Object {
                "contentTypes": Array [
                  "VIDEO",
                ],
              }
            }
            toggleItemSelection={[MockFunction]}
          />
        </div>
        <TableMobileDataToggleIcon
          isOpen={false}
          toggleDisplay={[Function]}
        >
          <Icon
            as="i"
            className="items_table_mobileDataToggleIcon"
            name="chevron down"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="chevron down icon items_table_mobileDataToggleIcon"
              onClick={[Function]}
            />
          </Icon>
        </TableMobileDataToggleIcon>
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="VISIBILITY"
      key="z132_visibility"
    >
      <td
        className="items_table_item"
        data-header="VISIBILITY"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            VISIBILITY
          </div>
          INTERNAL
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="MODIFIED"
      key="z132_updatedAt"
    >
      <td
        className="items_table_item"
        data-header="MODIFIED"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            MODIFIED
          </div>
          May 12, 2019
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="TEAM"
      key="z132_team"
    >
      <td
        className="items_table_item"
        data-header="TEAM"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            TEAM
          </div>
          IIP Video Production
        </span>
        <br />
      </td>
    </TableCell>
  </tr>,
  <tr
    className=""
  >
    <TableCell
      as="td"
      className="items_table_item"
      data-header="PROJECT TITLE"
      key="d137_projectTitle"
    >
      <td
        className="items_table_item"
        data-header="PROJECT TITLE"
      >
        <div
          className="primary_col"
        >
          <video-details-popup
            d={Object {}}
            header={
              Object {
                "label": "PROJECT TITLE",
                "name": "projectTitle",
              }
            }
            selectedItems={Array []}
            team={
              Object {
                "contentTypes": Array [
                  "VIDEO",
                ],
              }
            }
            toggleItemSelection={[MockFunction]}
          />
        </div>
        <TableMobileDataToggleIcon
          isOpen={false}
          toggleDisplay={[Function]}
        >
          <Icon
            as="i"
            className="items_table_mobileDataToggleIcon"
            name="chevron down"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="chevron down icon items_table_mobileDataToggleIcon"
              onClick={[Function]}
            />
          </Icon>
        </TableMobileDataToggleIcon>
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="VISIBILITY"
      key="d137_visibility"
    >
      <td
        className="items_table_item"
        data-header="VISIBILITY"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            VISIBILITY
          </div>
          INTERNAL
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="MODIFIED"
      key="d137_updatedAt"
    >
      <td
        className="items_table_item"
        data-header="MODIFIED"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            MODIFIED
          </div>
          May 9, 2019
        </span>
        <br />
      </td>
    </TableCell>
    <TableCell
      as="td"
      className="items_table_item"
      data-header="TEAM"
      key="d137_team"
    >
      <td
        className="items_table_item"
        data-header="TEAM"
      >
        <span>
          <div
            className="items_table_mobileHeader"
          >
            TEAM
          </div>
          IIP Video Production
        </span>
        <br />
      </td>
    </TableCell>
  </tr>,
]
`;
